File: ruby/task.rb
FILE_REMAINING = 'task.txt'.freeze
FILE_COMPLETED = 'completed.txt'.freeze

def help
  puts "Usage :-\n$ ./task add 2 hello world    # Add a new item with priority 2 and text \"hello world\" to the list\n$ ./task ls                   # Show incomplete priority list items sorted by priority in ascending order\n$ ./task del INDEX            # Delete the incomplete item with the given index\n$ ./task done INDEX           # Mark the incomplete item with the given index as complete\n$ ./task help                 # Show usage\n$ ./task report               # Statistics"
end

def get_file_data(which_file)
  data = nil
  begin
    File.open(which_file ? FILE_COMPLETED : FILE_REMAINING, 'r') do |fr|
      data = fr.readlines
    end
  rescue StandardError
    data = []
  end

  data - ['', nil, "\n"]
end

def set_file_data(data)
  File.open(FILE_REMAINING, 'w') do |fw|
    fw.puts(data)
  end
end

def get_remaining_tasks(which_file, display_quantity = false)
  data = get_file_data(which_file)

  puts 'There are no pending tasks!' if data.empty? && !display_quantity

  if display_quantity
    puts(which_file ? "Completed : #{data.length}" : "Pending : #{data.length}")
  end

  (0..data.length - 1).each do |i|
    item = data[i]
    item = item.strip
    if which_file
      puts "#{i + 1}. #{item}"
    else
      puts "#{i + 1}. #{item[item.index(' ') + 1, item.length]} [#{item[0, item.index(' ')]}]"
    end
  end
end

def set_new_task(priority, task)
  data = get_file_data(false)
  priority = priority.to_i
  puts 'Invalid index' unless priority
  ndx = 0
  data.each do |item|
    item = item.strip
    p = item[0, item.index(' ')].to_i
    break if priority < p

    ndx += 1
  end
  data.insert(ndx, "#{priority} #{task}")
  set_file_data(data)
  puts "Added task: \"#{task}\" with priority #{priority}"
end

def delete_task(index)
  index = index.to_i
  return nil unless index

  data = get_file_data(false)
  return nil if index < 1 || index > data.length

  deleted_item = data[index - 1]
  deleted_item = deleted_item[deleted_item.index(' ') + 1, deleted_item.length]
  data.delete_at(index - 1)
  set_file_data(data)
  deleted_item
end

def set_completed(index)
  deleted_item = delete_task(index)
  unless deleted_item
    puts "Error: no incomplete item with index ##{index} exists."
    return
  end

  File.open(FILE_COMPLETED, 'a') do |fa|
    fa.write(deleted_item)
  end
  puts 'Marked item as done.'
end

primary_arg = ARGV[0]
case primary_arg
when 'ls'
  get_remaining_tasks(false)
when 'add'
  if ARGV[1] && ARGV[2]
    set_new_task(ARGV[1], ARGV[2])
  else
    puts 'Error: Missing tasks string. Nothing added!'
  end

when 'del'
  if ARGV[1]
    if delete_task(ARGV[1])
      puts "Deleted task ##{ARGV[1]}"
    else
      puts "Error: task with index ##{ARGV[1]} does not exist. Nothing deleted."
    end
  else
    puts 'Error: Missing NUMBER for deleting tasks.'
  end

when 'done'
  if ARGV[1]
    set_completed(ARGV[1])
  else
    puts 'Error: Missing NUMBER for marking tasks as done.'
  end

when 'report'
  get_remaining_tasks(false, true)
  puts
  get_remaining_tasks(true, true)
else
  help
end

File: python/task.py
import sys

FILE_REMAINING = 'task.txt'
FILE_COMPLETED = 'completed.txt'
args = sys.argv

def help():
    print("Usage :-\n$ ./task add 2 hello world    # Add a new item with priority 2 and text \"hello world\" to the list\n$ ./task ls                   # Show incomplete priority list items sorted by priority in ascending order\n$ ./task del INDEX            # Delete the incomplete item with the given index\n$ ./task done INDEX           # Mark the incomplete item with the given index as complete\n$ ./task help                 # Show usage\n$ ./task report               # Statistics")


def getRemainingTasks(fileName, displayQuantity=None):
    try:
        with open(fileName, 'r') as fr:
            data = fr.readlines()
    except:
        data = []

    if not data and args[1] == 'ls':
        print('There are no pending tasks!')
        return

    if displayQuantity:
        if fileName == FILE_REMAINING:
            print('Pending :', len(data))
        else:
            print('Completed :', len(data))

    for i in range(len(data)):
        item = data[i]
        item = item.strip()
        if fileName == FILE_REMAINING:
            print('%s. %s [%s]' % (
                i+1, item[item.index(' ')+1:], item[0:item.index(' ')]))
        else:
            print('%s. %s' % (i+1, item))


def setNewTask(priority, task):
    try:
        with open(FILE_REMAINING, 'r') as fr:
            data = fr.readlines()
    except:
        data = []
    ndx = 0
    for item in data:
        if priority < item[0:item.index(' ')]:
            break
        ndx += 1
    if not data:
        fullTask = str(priority)+' '+task
    elif ndx == len(data):
        fullTask = '\n'+str(priority)+' '+task
    else:
        fullTask = str(priority)+' '+task+'\n'
    data.insert(ndx, fullTask)
    newData = ''.join(data)

    with open(FILE_REMAINING, 'w') as fr:
        fr.write(newData)
    print('Added task: "%s" with priority %s' % (task, priority))


def deleteTask(index):
    try:
        with open(FILE_REMAINING, 'r') as fr:
            data = fr.readlines()
    except:
        data = []

    try:
        index = int(index)
    except:
        print("Enter a valid index")
        return
    if index < 1 or index > len(data):
        return
    deletedItem = data[index-1]
    deletedItem = deletedItem[deletedItem.index(' ')+1:]
    del data[index-1]
    newData = ''.join(data)
    with open(FILE_REMAINING, 'w') as fr:
        fr.write(newData)
    return deletedItem


def setCompleted(index):
    deletedItem = deleteTask(index)
    if not deletedItem:
        print('Error: no incomplete item with index #%s exists.' % index, end='')
        return
    deletedItem = deletedItem.strip()
    with open(FILE_COMPLETED, 'a') as fc:
        fc.write(deletedItem+'\n')
    print("Marked item as done.")


primaryArg = None
if len(args) > 1:
    primaryArg = args[1]

if primaryArg == 'ls':
    getRemainingTasks(FILE_REMAINING)
elif primaryArg == 'add':
    if len(args) > 3:
        if args[3] >= '0':
            setNewTask(args[2], args[3])
        else:
            print('Priority cannot be negative')

    else:
        print("Error: Missing tasks string. Nothing added!")
elif primaryArg == 'del':
    if len(args) > 2:
        if deleteTask(args[2]):
            print('Deleted task #%s' % args[2], end='')
        else:
            print(
                'Error: task with index #%s does not exist. Nothing deleted.' % args[2], end='')

    else:
        print("Error: Missing NUMBER for deleting tasks.")

elif primaryArg == 'done':
    if len(args) > 2:
        setCompleted(args[2])
    else:
        print("Error: Missing NUMBER for marking tasks as done.")

elif primaryArg == 'report':
    getRemainingTasks(FILE_REMAINING, True)
    print()
    getRemainingTasks(FILE_COMPLETED, True)

else:
    help()
